File: Project-Examples/P0_Python/dal_layer/player_dao/player_dao_imp.py
from custom_exceptions.id_not_found import IdNotFound
from dal_layer.player_dao.player_dao_interface import PlayerDAOInterface
from entities.player_class_information import Player


class PlayerDAOImp(PlayerDAOInterface):
    player_list = [Player(1, 1, "The", "joker", 22)]
    id_generator = 2

    # Create
    def create_player(self, player: Player) -> Player:
        player.player_id = self.id_generator
        self.id_generator += 1
        self.player_list.append(player)
        return player

    # Read
    def get_player_by_id(self, player_id: int) -> Player:
        for player in self.player_list:
            if player.player_id == player_id:
                return player
        raise IdNotFound("No player matches the id given: please try again!")

    # Update
    def update_player_by_id(self, player: Player) -> Player:
        for old_player in self.player_list:
            if old_player.player_id == player.player_id:
                old_player = player
                return old_player
        raise IdNotFound("No player matches the id given: please try again!")

    # Delete
    def delete_player_by_id(self, player_id: int) -> bool:
        for player in self.player_list:
            if player.player_id == player_id:
                self.player_list.remove(player)
                return True
        raise IdNotFound("No player matches the id given: please try again!")

File: Week8/Day3/ConnectingToPostgres/src/main/java/com/revature/daos/PlayerDAOImp.java
package com.revature.daos;

import com.revature.entities.Player;
import com.revature.utils.DatabaseConnection;

import javax.xml.transform.Result;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class PlayerDAOImp implements PlayerDAO{

    @Override
    public Player createPlayer(Player player) {
        try(Connection connection = DatabaseConnection.createConnection()){
            // make our sql statement
            String sql = "insert into players values(default, ?, ?)";
            // create a prepared statement object
            PreparedStatement ps = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS);
            // use ps object to execute our query and pass in necessary paramenters
            ps.setString(1, player.getFirstName());
            ps.setString(2, player.getLastName());
            ps.execute();
            // create a result set from our query
            ResultSet rs = ps.getGeneratedKeys();
            // retrieve info from rs
            rs.next();
            player.setPlayerId(rs.getInt("player_id"));
            // return updated player object
            return player;
        } catch(SQLException e){
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public Player selectPlayerById(int id) {
        try(Connection connection = DatabaseConnection.createConnection()){
            String sql = "select * from players where player_id = ?";
            PreparedStatement ps = connection.prepareStatement(sql);
            ps.setInt(1,id);
            ResultSet rs = ps.executeQuery(); // executeQuery returns a result set object with the results
            rs.next();
            Player player = new Player(
                    rs.getInt("player_id"),
                    rs.getString("first_name"),
                    rs.getString("last_name")
            );
            return player;
        } catch(SQLException e){
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public List<Player> selectAllPlayers() {
        try(Connection connection = DatabaseConnection.createConnection()){
            String sql = "select * from players";
            // can create a statement, not prepared statement, if there are no parameters to add
            Statement s = connection.createStatement();
            s.execute(sql);
            ResultSet rs = s.getResultSet();
            List<Player> players = new ArrayList<>();
            while(rs.next()){
                Player player = new Player(
                        rs.getInt("player_id"),
                        rs.getString("first_name"),
                        rs.getString("last_name")
                );
                players.add(player);
            }
            return players;
        } catch(SQLException e){
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public int updatePlayerById(Player player) {
        try(Connection connection = DatabaseConnection.createConnection()){
            String sql = "update players set first_name = ?, last_name = ? where player_id = ?";
            PreparedStatement ps = connection.prepareStatement(sql);
            ps.setString(1, player.getFirstName());
            ps.setString(2, player.getLastName());
            ps.setInt(3, player.getPlayerId());
            return ps.executeUpdate();
        } catch(SQLException e){
            e.printStackTrace();
            return 0;
        }
    }

    @Override
    public int deletePlayerById(int id) {
        try(Connection connection = DatabaseConnection.createConnection()){
            String sql = "delete from players where player_id = ?";
            PreparedStatement ps = connection.prepareStatement(sql);
            ps.setInt(1,id);
            return ps.executeUpdate();
        } catch(SQLException e){
            e.printStackTrace();
            return 0;
        }
    }
}
